### ISSUE TITLE
Deprecate non-AMP block on consent API

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#8275]
## Feature Description

In https://github.com/google/site-kit-wp/issues/2087 we introduced a simple API for implementing manual disabling of tags based for web and AMP. AMP has built-in support for blocking elements based on consent, so a manual approach was adopted for web tags.

With consent mode, this should be deprecated to encourage site owners to use consent mode instead.

**This should only take effect when the `consentMode` feature is enabled.**

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* ~The existent manual disabling of the web tag should be retained for Analytics but only activated when Consent Mode is disabled.~

Please note that the AC for this issue has been updated as follows:

* Usage of the `googlesitekit_analytics-4_tag_block_on_consent` filter should be deprecated.
* When a hook is run for this filter, a deprecation warning should be logged to the debug log.
* Aside from the deprecation warning, the legacy manual disabling of the web tag should still function as usual.

## Implementation Brief

* [x] Extract this block of code that disables the web tag into a new method.
https://github.com/google/site-kit-wp/blob/28780f2a4a3502573d513eb1a82a931284b874fd/includes/Modules/Analytics_4/Web_Tag.php#L161-L176
* [x] Mark the new method as deprecated via a calls to `_deprecated_function()`.
* [x] Conditionally invoke the method when the Consent Mode `enabled` setting is `false`.

### Test Coverage

* Provide PHPUnit test coverage for the above changes.

### QA:Eng

- Copy the following PHP snippet to a new `.php` file in `wp-content/mu-plugins/` to enable legacy tag blocking:
```php
<?php
add_filter( 'googlesitekit_analytics-4_tag_block_on_consent', '__return_true' );
```
- Look for the `type="text/plain" data-block-on-consent` attributes on the GTag script to confirm legacy tag blocking is enabled:
```html
<!-- Google Analytics snippet added by Site Kit -->
<script type="text/plain" data-block-on-consent src='https://www.googletagmanager.com/gtag/js?id=GT-ABC123' id='google_gtagjs-js' async></script>
```
- Verify that a warning is raised when the deprecated filter `googlesitekit_analytics-4_tag_block_on_consent` is used, while other `googlesitekit_{module_slug}_tag_block_on_consent` filters don't raise a warning.
- Ensure the behaviour related to the above filters still works as expected.

## Changelog entry

* Deprecate legacy web tag block_on_consent for Analytics when Consent Mode is enabled.

**Comment by aaemnnosttv** [comment]
AC ✅ 

IB

> * Add `is_tag_blocked_on_consent` getter/setter methods to the `GTag` class.

The tag modifications that this would control don't necessarily need to be part of the GTag class since it filters the output via `script_loader_tag`, but we could keep it there.

> * Set the above to `true` in `Analytics_4/Web_Tag` if `$this->get_tag_blocked_on_consent_attribute()` returns `true`.

This doesn't return `true`, but a string of attributes if not `false`.

> Mark the above as deprecated in the code.

Let's call `wc_deprecated_hook` for such cases to notify that consent mode should be used instead. I was thinking we'd do this with `apply_filters_deprecated` from `Module_Web_Tag::get_tag_blocked_on_consent_attribute` but there are other web tags (like AdSense) that don't use consent mode, so we'd want to be more specific.

**Comment by techanvil** [comment]
> > * Add `is_tag_blocked_on_consent` getter/setter methods to the `GTag` class.
> 
> The tag modifications that this would control don't necessarily need to be part of the GTag class since it filters the output via `script_loader_tag`, but we could keep it there.

Good point; I figured it _might_ be reusable and it centralises the check for the Consent Mode setting, but we could simply do the same in `Web_Tag`. Anyhow I've kept it in `GTag`, it seems OK.

> > * Set the above to `true` in `Analytics_4/Web_Tag` if `$this->get_tag_blocked_on_consent_attribute()` returns `true`.
> 
> This doesn't return `true`, but a string of attributes if not `false`.

Thanks for pointing this out - I did notice this while digging into the code, but forgot when I was drafting the IB, which I've amended to fix.

> > Mark the above as deprecated in the code.
> 
> Let's call `wc_deprecated_hook` for such cases to notify that consent mode should be used instead. I was thinking we'd do this with `apply_filters_deprecated` from `Module_Web_Tag::get_tag_blocked_on_consent_attribute` but there are other web tags (like AdSense) that don't use consent mode, so we'd want to be more specific.

Did you mean `_deprecated_hook()` ? Seeing as it's not a hook per se I've opted for a different approach, adding calls to `_deprecated_function()` to the getter/setters. Happy to change this as you prefer, though.

**Comment by techanvil** [comment]
Update: I've realised this may need to change depending on the outcome of whether we keep https://github.com/google/site-kit-wp/issues/8269 in this epic, so moving this back to AC for now.

**Comment by techanvil** [comment]
Following the decision to remove #8269 from this epic, I've updated the AC and IB for this issue.

**Comment by techanvil** [comment]
This seems pretty straightforward, so I am moving it to execution.

**Comment by bethanylang** [comment]
@techanvil Is this still blocked by #8273 or can that dependency be removed?

**Comment by techanvil** [comment]
@bethanylang thanks for spotting that, this issue is no longer blocked by #8273 and I've removed the dependency.

**Comment by techanvil** [comment]
@mohitwp as mentioned on Slack I've moved this back to Execution for a quick followup PR to reduce the number of deprecation warnings this issue has caused.

**Comment by mohitwp** [comment]
Moving this to execution as per Slack conversation.

**Comment by mohitwp** [comment]
### QA Update ⚠️ 

@techanvil 

- Tested on dev environmet.
- Verified tag blocking is only  enabled when
      - Enabled via the PHP snippet AND
      -  The consentMode feature flag is disabled OR the consentMode feature flag is enabled while the Consent Mode enabled setting is false.
- Also, verified that Tag blocking is not enabled via PHP then `type="text/plain" data-block-on-consent` `attributes on the GTag script is not showing. 

<details>

![image](https://github.com/google/site-kit-wp/assets/94359491/dd688c9d-1e36-44a3-adee-8f2b8612365f)

</details>

**_Question-1)_** When I enabled the tag blocking through PHP snippet then in Front end I'm getting this deprecated error - `Deprecated:  Function Google\Site_Kit\Modules\Analytics_4\Web_Tag::add_legacy_block_on_consent_attributes is deprecated since version n.e.x.t with no alternative available. in /s3-rightfulmeal/wordpress/wp-includes/functions.php on line 6031
`

<details>

![image](https://github.com/google/site-kit-wp/assets/94359491/ad5aa964-53b2-4ddd-b386-14fb426317cf)

</details>

**_Question 2_** : Tag blocking is also enabled when

_Scenario 1_
-  Enabled via the PHP snippet AND
-  the consentMode feature flag is enabled while the Consent Mode enabled setting is true.

<details>

![image](https://github.com/google/site-kit-wp/assets/94359491/df7320e2-954f-4fd2-a79c-0c0399267197)

![image](https://github.com/google/site-kit-wp/assets/94359491/2d34bc92-7e55-4b12-bc9b-0c07a3325f3f)

</details>

_Scenario 2_
-  Enabled via the PHP snippet AND
-  the consentMode feature flag is disabled while the Consent Mode enabled setting is true.

<details>

![image](https://github.com/google/site-kit-wp/assets/94359491/cf5f77ec-dc37-430c-9635-582f7ca716cf)

![image](https://github.com/google/site-kit-wp/assets/94359491/e587b4b1-edb0-4323-96ba-455d3e4fce90)

</details>

 **Is this expected ?**

**Comment by techanvil** [comment]
> **_Question-1)_** When I enabled the tag blocking through PHP snippet then in Front end I'm getting this deprecated error - `Deprecated: Function Google\Site_Kit\Modules\Analytics_4\Web_Tag::add_legacy_block_on_consent_attributes is deprecated since version n.e.x.t with no alternative available. in /s3-rightfulmeal/wordpress/wp-includes/functions.php on line 6031 `

As discussed [here](https://github.com/google/site-kit-wp/issues/8274#issuecomment-1966665124) this indicates you've got `WP_DEBUG` etc enabled, which we wouldn't expect to see on a production site, and it's still possible to suppress logging these warnings to the front-end via `WP_DEBUG_DISPLAY`. We've reduced the deprecation warnings so they only occur when a user actually has the deprecated feature in use, so I think we do want to keep them. Therefore this is behaving as expected.

> **_Question 2_** : Tag blocking is also enabled when
> 
> _Scenario 1_
> 
> * Enabled via the PHP snippet AND
> * the consentMode feature flag is enabled while the Consent Mode enabled setting is true.

We should _not_ expect to see the legacy attributes in this scenario. Did you definitely refresh the page after enabling Consent Mode?

> _Scenario 2_
> 
> * Enabled via the PHP snippet AND
> * the consentMode feature flag is disabled while the Consent Mode enabled setting is true.

We _should_ expect to see the legacy attributes in this scenario as the feature flag is disabled.

**Comment by mohitwp** [comment]
### QA Update ✅ 

Thank you @techanvil !
I'm getting expected results for all the scenarios. 

- Tested on dev environmet.
- Verified tag blocking is only  enabled when
      - Enabled via the PHP snippet AND
      -  The consentMode feature flag is disabled OR the consentMode feature flag is enabled while the Consent Mode enabled setting is false.
- Also, verified that Tag blocking is not enabled via PHP then `type="text/plain" data-block-on-consent` `attributes on the GTag script is not showing. 

<details>

![image](https://github.com/google/site-kit-wp/assets/94359491/dd688c9d-1e36-44a3-adee-8f2b8612365f)

</details>

Also, verified below listed scenarios -

_Scenario 1_
-  Enabled via the PHP snippet AND
-  the consentMode feature flag is enabled while the Consent Mode enabled setting is true.

**Result** - Tag blocking is not enabled. 

_Scenario 2_
-  Enabled via the PHP snippet AND
-  the consentMode feature flag is disabled while the Consent Mode enabled setting is true.

**Result** - Tag blocking is enabled.

<details>

![image](https://github.com/google/site-kit-wp/assets/94359491/cf5f77ec-dc37-430c-9635-582f7ca716cf)

![image](https://github.com/google/site-kit-wp/assets/94359491/e587b4b1-edb0-4323-96ba-455d3e4fce90)

</details>

**Comment by aaemnnosttv** [comment]
As noted on the call earlier, we need to update the deprecation warning to be specific to the filter rather than our internal function call. This shouldn't invalidate the QA that's been done so far but let's confirm once updated.

**Comment by aaemnnosttv** [comment]
@techanvil I've created a POC for what could work as a follow-up here: https://github.com/google/site-kit-wp/pull/8359

In doing so, I also found that the previous logic which only triggered the deprecated hook notice was tied to consent mode being enabled which wasn't entirely accurate either but I understand why it was done. The notice should be emitted if the legacy block on consent filter is used (at all) but only for GA, which is what my POC does. A key detail is how `apply_filters_deprecated` works which only triggers the notice if a callback is added on the filter, otherwise it works as normal (meaning nothing was hooked on it).

This allowed some tests to be simplified a little for GA4 and the removal of "set_consent_mode_enabled" parts on the GA4 web tag.

I'm open to taking a simpler approach for the sake of time if it makes sense, but I think there may have been a misunderstanding due to the way `_deprecated_function` was used before.

I'll assign this back over to you to look over and iterate on if necessary.

**Comment by techanvil** [comment]
Thanks @aaemnnosttv - yes, I got a bit confused as to the intended approach here. Your POC LGTM, I have merged it, although I didn't notice we still needed to remove the `_deprecated_function()` call, so have raised a quick followup for that.

FAO the code reviewer for this followup, as mentioned on Slack I was in the middle of doing a `QA:Eng` test to verify the deprecated filter behaviour when I realised this additional followup was needed, so please assign it back to me in QA so I can finish this off.

**Comment by aaemnnosttv** [comment]
> Your POC LGTM, I have merged it, although I didn't notice we still needed to remove the `_deprecated_function()` call, so have raised a quick followup for that.

Nice catch, I thought I removed that 👍 

Looks like there was some of the removed code that was added back while merging in main so I've fixed that as well as removing a related condition that was missed https://github.com/google/site-kit-wp/blob/b687a1095f529c68990a959d01651a517a53a2ba/includes/Modules/Analytics_4/Web_Tag.php#L184-L190

Finally, I removed the `deprecated` annotations on our internal methods since those aren't deprecated, only the use of the legacy API is. We need to think a bit more about how we move to some web tags still using it while others won't since not all tags support consent mode.

Once things finish running I'll merge this and send it back to you for QA 👍  Thanks!

**Comment by techanvil** [comment]
## QA :white_check_mark: 

- Connected Analytics:

![image](https://github.com/google/site-kit-wp/assets/18395600/893e6006-336a-4cf0-abad-732df3a83eb4)

- Verified the debug log was empty:

![image](https://github.com/google/site-kit-wp/assets/18395600/aee5515d-a7ad-4845-80fa-9c79c2b50a87)

- Enabled the `block_on_consent` filter for `analytics-4`:
  - Note, for some reason `mu-plugins` wasn't working on my InstaWP test site so I added the filter in `functions.php` instead.

![image](https://github.com/google/site-kit-wp/assets/18395600/5ee139db-7784-406b-8149-3977e966647b)

- Verified the `type="text/plain"` and `data-block-on-consent` attributes appeared as expected:

![image](https://github.com/google/site-kit-wp/assets/18395600/62795395-922b-4ab3-b7ff-4dff1d1cff75)

- Verified the deprecated hook error appeared in the debug log:

![image](https://github.com/google/site-kit-wp/assets/18395600/73f191fc-d381-4fbf-8524-6699af92b5c8)

- Connected AdSense to verify the non-Analytics path:

![image](https://github.com/google/site-kit-wp/assets/18395600/79a9b4ab-4a66-4bd1-818c-51dbcc3a1013)

- Enabled the `block_on_consent` filter for `analytics-4`:

![image](https://github.com/google/site-kit-wp/assets/18395600/e5032a96-3345-485a-baf2-de62093dfebd)

- Verified the `type="text/plain"` and `data-block-on-consent` attributes appeared as expected:

![image](https://github.com/google/site-kit-wp/assets/18395600/fb73d0b3-9bbd-44a9-a967-ea90804182f2)

- Verified that a deprecated hook error did not appear for `adsense` in the debug log:

![image](https://github.com/google/site-kit-wp/assets/18395600/9ac1a113-5428-44c9-9bc0-834ea7b20ad5)

- I ran through this with Consent Mode both disabled and enabled with the same result.
